src/ILCompiler.Compiler/src/IL/ILImporter.Scanner.cs

```diff
@@ -111,8 +111,13 @@ public DependencyList Import()
                 if (!_isFallbackBodyCompilation &&
                     (_canonMethod.Signature.IsStatic || _canonMethod.IsConstructor || owningType.IsValueType))
                 {
-                    // For beforefieldinit, we can wait for field access.
-                    if (!((MetadataType)owningType).IsBeforeFieldInit)
+                    // For beforefieldinit, we can wait for field access...
+                    bool runCctor = !((MetadataType)owningType).IsBeforeFieldInit;
+
+                    // ...unless this is a PInvoke, in which case we run it regardless
```

Comment:
> I don't think it should for the program above.

I agree that it should not for the program above. It does not trigger the cctor in .NET Framework or CoreCLR either (when the program is compiled `/o+`).

The RyuJIT calls `initClass` method on the JIT/EE interface to decide whether to emit the cctor trigger. There maybe a bug in this method. I would be useful to understand.